// @backstage/backend-test-utils/dist/index.cjs.js
'use strict';

var TestCaches = require('./cache/TestCaches.cjs.js');
var TestDatabases = require('./database/TestDatabases.cjs.js');
var registerMswTestHooks = require('./msw/registerMswTestHooks.cjs.js');
var MockDirectory = require('./filesystem/MockDirectory.cjs.js');
var ServiceFactoryTester = require('./next/wiring/ServiceFactoryTester.cjs.js');
var TestBackend = require('./next/wiring/TestBackend.cjs.js');
var mockServices = require('./next/services/mockServices.cjs.js');
var mockCredentials = require('./next/services/mockCredentials.cjs.js');
var errorHandler = require('./util/errorHandler.cjs.js');



exports.TestCaches = TestCaches.TestCaches;
exports.TestDatabases = TestDatabases.TestDatabases;
exports.registerMswTestHooks = registerMswTestHooks.registerMswTestHooks;
exports.createMockDirectory = MockDirectory.createMockDirectory;
exports.ServiceFactoryTester = ServiceFactoryTester.ServiceFactoryTester;
exports.startTestBackend = TestBackend.startTestBackend;
Object.defineProperty(exports, "mockServices", {
	enumerable: true,
	get: function () { return mockServices.mockServices; }
});
Object.defineProperty(exports, "mockCredentials", {
	enumerable: true,
	get: function () { return mockCredentials.mockCredentials; }
});
exports.mockErrorHandler = errorHandler.mockErrorHandler;
