// @backstage/plugin-scaffolder-node-test-utils/dist/actions/mockActionContext.cjs.js
'use strict';

var stream = require('stream');
var backendCommon = require('@backstage/backend-common');
var backendTestUtils = require('@backstage/backend-test-utils');

const createMockActionContext = (options) => {
  const credentials = backendTestUtils.mockCredentials.user();
  const defaultContext = {
    logger: backendCommon.loggerToWinstonLogger(backendTestUtils.mockServices.logger.mock()),
    logStream: new stream.PassThrough(),
    output: jest.fn(),
    createTemporaryDirectory: jest.fn(),
    input: {},
    checkpoint: jest.fn(),
    getInitiatorCredentials: () => Promise.resolve(credentials)
  };
  const createDefaultWorkspace = () => ({
    workspacePath: backendTestUtils.createMockDirectory().resolve("workspace")
  });
  if (!options) {
    return {
      ...defaultContext,
      ...createDefaultWorkspace()
    };
  }
  const { input, logger, logStream, secrets, templateInfo, workspacePath } = options;
  return {
    ...defaultContext,
    ...workspacePath ? { workspacePath } : createDefaultWorkspace(),
    ...workspacePath && {
      createTemporaryDirectory: jest.fn().mockResolvedValue(workspacePath)
    },
    ...logger && { logger },
    ...logStream && { logStream },
    ...input && { input },
    ...secrets && { secrets },
    templateInfo
  };
};

exports.createMockActionContext = createMockActionContext;
